src/quil.rs

```diff
@@ -0,0 +1,29 @@
+/// A trait to wrap items which represent some construct within the Quil language.
+pub trait Quil {
+    type Display: std::fmt::Display;
+
+    /// Return a string in valid Quil syntax or an error if the item cannot be represented with valid Quil.
+    fn to_quil(&self) -> Result<Self::Display, ToQuilError>;
+
+    /// Return a string in valid Quil syntax if possible or otherwise the debug representation of the item.
+    fn to_quil_or_debug(&self) -> String
+    where
+        Self: std::fmt::Debug,
+    {
+        self.to_quil()
+            .map(|s| s.to_string())
+            .unwrap_or_else(|_| format!("{:#?}", self))
+    }
+}
```

Comment:
I'm curious if this is as efficient as it could be? The implementation of `to_quil` looks like it often requires the allocation of a new strings. Maybe trivial for simple implementations, but a lot of instructions are repeatedly calling `format!` in an iterator.

Would something like this would work?

```rs
pub trait Quil {
    fn write(&self, w: &mut String) -> Result<(), ToQuilError>;

    fn try_to_quil(&self) -> Result<String, DisplayError> {
        let mut w = String::new();
        self.write(&mut w)?;
        Ok(w)
    }
}
```

Then, implementations can take advantage of `write!`, writing new data to a buffer instead of allocating and concatenating new strings each time `format!` needs to be called.

```rs
use std::fmt::write;

impl Quil for Instruction {
    fn write(&self, w: &mut String) -> Result<(), ToQuilError> {
        write!(w, "{}", self.value)
    }
}
```

Overall, this seems clearer, and easier to implement, as it doesn't require a custom type definition, and it pretty much looks and works exactly like `Display::fmt`.

Reply:
Thinking on this a bit more, the argument to `write` could probably be something like `impl Write` to avoid needing to import the trait and to decouple the buffer type from implementations.

Reply:
Yep, that could work! This is the kind of thing I was looking for in our group chat earlier. I'll keep `to_quil`, I think, but give it a default implementation that defers to `write`. What do you think about `write_quil` in place of `write`?